apm-sniffer/apm-sdk-plugin/grpc-1.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/grpc/v1/server/TracingServerCallListener.java

```diff
@@ -101,4 +96,18 @@ public void onHalfClose() {
             ContextManager.stopSpan();
         }
     }
+
+    @Override
+    public void onComplete() {
+        super.onComplete();
+        ContextManager.stopSpan();
```

Comment:
Is this always in the same thread?

Reply:
in the same thread

Reply:
give me an email address?i will describe detail  to you what problem we met

Reply:
All case should be open and repeated. You could write the case and graph here.

Reply:
in the class`org.apache.skywalking.apm.plugin.grpc.v1.server.ServerInterceptor` will create an entry gRPC endpoint,but it always cost a few milliseconds,because it just new a `TracingServerCallListener`.
```
final AbstractSpan span = ContextManager.createEntrySpan(OperationNameFormatUtil.formatOperationName(call.getMethodDescriptor()), contextCarrier);
        span.setComponent(ComponentsDefine.GRPC);
        span.setLayer(SpanLayer.RPC_FRAMEWORK);
        try {
            return new TracingServerCallListener<>(handler.startCall(new TracingServerCall<>(call, ContextManager.capture()), headers), call.getMethodDescriptor(), ContextManager.capture());
        } finally {
            ContextManager.stopSpan();
        }

```
so ,it cause  the avg response of gRPC is nearly 0 ms.
![image](https://user-images.githubusercontent.com/11434590/123513648-bf828900-d6c0-11eb-9230-d4431a7b323e.png)

the really cost some time is the endpoint of xx/server/Request/onComplete,

![image](https://user-images.githubusercontent.com/11434590/123513741-59e2cc80-d6c1-11eb-8b58-2c27123c98b0.png)

`org.apache.skywalking.apm.plugin.grpc.v1.server.TracingServerCallListener`
```
@Override
    public void onHalfClose() {
        final AbstractSpan span = ContextManager.createLocalSpan(operationPrefix + REQUEST_ON_COMPLETE_OPERATION_NAME);
        span.setComponent(ComponentsDefine.GRPC);
        span.setLayer(SpanLayer.RPC_FRAMEWORK);
        ContextManager.continued(contextSnapshot);

        try {
            super.onHalfClose();
        } catch (Throwable t) {
            ContextManager.activeSpan().errorOccurred().log(t);
            throw t;
        } finally {
            ContextManager.stopSpan();
        }
    }
```

Reply:
I agree that, `onComplete` or `onError` should be the right place to finish span. Only concern is about across threads.

Reply:
![image](https://user-images.githubusercontent.com/11434590/123516642-ba790600-d6cf-11eb-8f06-a191ea562dbe.png)
I print the thread id in the corresponding location ,thread is all the same
```
@Override
    public void onComplete() {
        String log ="method:%s,threadId:%s,currentTimeMillis:%s";
        System.out.println(String.format(log,"TracingServerCallListener.onComplete", Thread.currentThread().getId(), System.currentTimeMillis()));
        super.onComplete();
    }

    @Override
    public void onReady() {
        String log ="method:%s,threadId:%s,currentTimeMillis:%s";
        System.out.println(String.format(log,"TracingServerCallListener.onReady", Thread.currentThread().getId(), System.currentTimeMillis()));
        super.onReady();
    }
```

Is there anything I haven't considered

Reply:
> Is there anything I haven't considered

I am not sure how it looks like in high traffic load. Thread scheduling changes with load.

Reply:
so you worry about the onReady method execution thread is not the same as the thread when executing onComplete

`org.apache.skywalking.apm.plugin.grpc.v1.server.TracingServerCallListener`
```
@Override
    public void onComplete() {
        super.onComplete();
        ContextManager.stopSpan();
    }

    @Override
    public void onReady() {
        final AbstractSpan span = ContextManager.createEntrySpan(operation, contextCarrier);
        span.setComponent(ComponentsDefine.GRPC);
        span.setLayer(SpanLayer.RPC_FRAMEWORK);
        super.onReady();
    }
```

Reply:
As you are using `ContextManager#activeSpan`, then once they are not in the same thread, the memory leaks.

If there is a chance they are not(I didn't read gRPC source codes) always in the same thread, you should make the created in async mode, and use `span#asyncFinish` to close it safely. Of course, only need to do so if it crosses threads in some cases.